README for batch-processing project (#49)
CodeYourFuture/immersive-go-course: batch-processing/README.md

# Batch Processing

> In this project, you'll build a simple image processing pipeline: reading from a CSV file, downloading an image, processing it, and uploading it to cloud storage.

Reviewer: ```suggestion
In this project, you'll build a simple image processing pipeline: reading a list of image URLs from a CSV file, downloading each image, processing it, and uploading it to cloud storage.
```

> ⚠️ This project requires you to have access to an Amazon AWS account, with permissions to configure IAM and S3. Ask on CYF Slack for help with that.

> ⚠️ You need a GitHub repo to complete this because we'll be using [GitHub Actions](https://docs.github.com/en/actions).

> Learning objectives:

Reviewer: I'd also maybe add an LO (which I really like that we're doing) which is how to read/modify/extend existing code. Super common software skill which I'm not sure they'll have really done much before. I really like that this is doing that!

- What is batch processing and how does it differ from building servers?

> - How do we build resilience into batch processing software?

Reviewer: I'm not sure we've quite gotten this one yet :)

Author: No indeed.

Author: I've now added some pointers & extensions. We can return to this project later to tweak and improve

- How do we use Go to use run existing software to complete tasks?
- What is cloud storage technology and how do we read from it & upload data to it?
<!-- * How do we deploy batch processing tasks in the cloud? -->

Timebox: 2 days

## Project

This project starts with some scaffolding code already written. We'll be using a tool called [ImageMagick](https://imagemagick.org/) to process the images. The specific details of it are not important, so this is provided for us.

ImageMagick is software written in C. We're going to call out to this code from Go using a package called [imagick](https://github.com/gographics/imagick), which acts an an interface between our Go code and the ImageMagick code.

As a result of this dependency on ImageMagick, we're going to run this whole project in Docker. That means, rather than use `go run` or `go test`, we'll build and run a Docker container that has ImageMagick correctly installed.

> > It's possible to install ImageMagick on your computer to run this, but you don't _need_ to do that.

Reviewer: ```suggestion
> It's possible to install ImageMagick on your computer to run this outside of docker, but you don't _need_ to do that.
```

The Docker commands get a bit complex so we can do things like run tests and _see_ the output of the code, so we're not going to type the Docker commands ourselves.

> Instead, to run the code, we use a tool called [Make](https://www.gnu.org/software/make/). Make enables us to build code without knowing the details of how that is done, because these details are recorded in the `Makefile`. In that way it's similar to Docker: the complexity of how to run the code in a container is supplied in the `Dockerfile`.

Reviewer: ```suggestion
Instead, to run the code, we use a tool called [Make](https://www.gnu.org/software/make/). Make enables us to build code without knowing the details of how that is done, because these details are recorded in the `Makefile`. In that way it's similar to Docker - when using Make you store the complexity of what to run in a `Makefile`, and when using Docker you store the complexity of how to run the code in a container in a `Dockerfile`.
```

I think there's quite a lot of nuance here; I'd maybe consider even just removing the comparison to Docker... Unsure...

> Read a [quick introduction to Make](https://www.gnu.org/software/make/) and then install it for your computer. On macOS, it's likely that you have it already — try typing `make` into your termincal. If not, you need to [install the command-line tools](https://www.freecodecamp.org/news/install-xcode-command-line-tools/).

Reviewer: ```suggestion
Read a [quick introduction to Make](https://www.gnu.org/software/make/) and then install it for your computer. On macOS, it's likely that you have it already — try typing `make` into your terminal. It should output something like `make: no target specified`. If not (e.g. it says something like `command not found: make`, you need to [install the command-line tools](https://www.freecodecamp.org/news/install-xcode-command-line-tools/).
```

> > **Important**: you don't need to understand `make` in depth. We're only going to use the very simplest features.

Reviewer: ```suggestion
> **Important**: you don't need to understand `make` in depth. We're only going to use the very simplest features. In particular, you shouldn't have to edit the `Makefile` we've supplied other than exactly as we describe (but feel free if you want to!)
```

 => => writing image sha256:7f24030e5dc71ed23f98885644321c76b3d6709f7140391940ff06f486a33372         0.0s
 => => naming to docker.io/library/test                                                              0.0s

docker run \
		--rm test
=== RUN   TestGrayscaleMockError
--- PASS: TestGrayscaleMockError (0.00s)
=== RUN   TestGrayscaleMockCall
--- PASS: TestGrayscaleMockCall (0.00s)
PASS
ok  	github.com/CodeYourFuture/immersive-go-course/batch-processing	0.011s
```

This will have run much faster, because all the steps are cached by Docker. This is called the [build cache](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#leverage-build-cache).

Neat! We've now got the application and the tests running in Docker.

## Specificiation

So now we've got some running code, let's get started.

We want a **CLI tool** that:

- reads an input CSV containing URLs
- downloads images by URL
- processes them using ImageMagick to make them monochrome/grayscale
- uploads the results to [Amazon AWS S3 cloud storage](https://aws.amazon.com/s3/)
- writes an output CSV describing what it did

> It should run like this: `go run . --input input.csv --output output.csv`

Reviewer: ```suggestion
It should run like this: `go run . --input input.csv --output output.csv`

You probably want to modify the `main.go` file we've supplied in this directory, re-using some bits of it.
```

I'm not sure this is otherwise clear.

The basic organisation of S3 is simple. At the top level of S3 is the **bucket**. Data ("objects") are stored within these buckets in folders, like a normal file system. Read the [high-level guide to S3](https://docs.aws.amazon.com/AmazonS3/latest/userguide/Welcome.html).

You will need to create a bucket — call it anything you like. Pay attention to which [AWS region](https://cloudacademy.com/blog/aws-regions-and-availability-zones-the-simplest-explanation-you-will-ever-find-around/#:~:text=What%20are%20AWS%20Regions%3F,host%20their%20cloud%20infrastructure%20there.) you create it in – you will need this later.

You can make it publically accessible to the internet, so that you or anyone can view the images that are uploaded, with a policy like this:

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "PublicAccess",
      "Effect": "Allow",
      "Principal": "*",
      "Action": ["s3:GetObject", "s3:GetObjectVersion"],
      "Resource": "arn:aws:s3:::[NAME OF BUCKET]/*"
    }
  ]
}
```

> > ⚠️ Public means public. **DO NOT** upload anything sensitive to the bucket.

Reviewer: Good call-out :)

Use the [AWS SDK for Go](https://github.com/aws/aws-sdk-go) to interact with Amazon S3.

You will need to set up credentials so that your docker image can write to S3. The best way to do that is to:

- Create a role + policy to allow uploads to the bucket
- Allow your AWS user account to use the role

> - Make your AWS credentials available to the Docker image

Reviewer: ```suggestion
- Make your AWS credentials available to the Docker container
```

I'd generally lean towards "mount secrets at runtime" rather than "bake secrets into an image" :)

Author: Ooft, yes. This was my intent, wrong word. Thanks.

#### Role + policy

[Create a Policy](https://us-east-1.console.aws.amazon.com/iamv2/home#/policies) `S3ReadWriteGoCourse` with the following configuration, which allows reading and writing to the bucket.

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "ListObjectsInBucket",
      "Effect": "Allow",
      "Action": ["s3:ListBucket"],
      "Resource": ["arn:aws:s3:::[NAME OF BUCKET]"]
    },
    {
      "Sid": "AllObjectActions",
      "Effect": "Allow",
      "Action": "s3:*Object",
      "Resource": ["arn:aws:s3:::[NAME OF BUCKET]/*"]
    }
  ]
}
```

[Create a Role](https://us-east-1.console.aws.amazon.com/iamv2/home#/roles) `GoCourseLambdaUserReadWriteS3` associated with the `S3ReadWriteGoCourse` Policy. Set up the Trust Relationships of this policy as follows:

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "AWS": "arn:aws:iam::[YOUR AWS USER ID]:root"
      },
      "Action": "sts:AssumeRole"
    },
    {
      "Effect": "Allow",
      "Principal": {
        "Service": "lambda.amazonaws.com"
      },
      "Action": "sts:AssumeRole"
    }
  ]
}
```

> One that's done, you need local credentials:

Reviewer: ```suggestion
Once that's done, you need local credentials:
```

1. Install the [AWS CLI](https://aws.amazon.com/cli/)
1. [Follow this guide](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-quickstart.html) to set up your credentials locally

> Then change the `Makefile` to [mount](https://docs.docker.com/storage/bind-mounts/) your `.aws` directory into the Docker container, in the `root` user's home directory. When using the Go SDK, it will automatically find these credentials and use them:

Reviewer: ```suggestion
Then change the `Makefile` to [mount](https://docs.docker.com/storage/bind-mounts/) your `.aws` directory into the Docker container, in the `root` user's home directory (`/root/.aws`). When using the Go SDK, it will automatically find these credentials and use them:
```

```Makefile
run: outputs
	docker ...
        ...
        --mount type=bind,source="$$(echo $$HOME)/.aws",target=/root/.aws \
		...
```

You'll need to say which AWS region the S3 bucket is located in. Do that with the `AWS_REGION` environment variable.

> In your `Dockerfile`:

Reviewer: Do you actually need to declare this in the `Dockerfile`? I think `-e` should work even if not?

Author: Perhaps not! In some iteration it had a value here.
